### PR TITLE
PHP-DI 7.0

### PR DESCRIPTION
**Pull request by mnapoli**
This pull request contains a possible v7.0 version.

Changes:

- support PHP 8.0 and up
- #769 Allow `new Container()` with definitions
- #738, #741 support PHP 8 attributes for injections: welcome `#[Inject]` to replace `@Inject`!
- drop support for `@Inject` used with phpdoc annotations -> only native PHP types are supported now
    That makes PHP-DI simpler, faster and lighter (no need for the PhpDocReader package)
- typed everything that can be typed
- PSR-11 1.2 and 2.0 support

TODO:

- [x] #718 
- [x] #812 

Blocked by:

- [x] https://github.com/opis/closure/pull/67
- [x] https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4702

### CODE REVIEW

**Comment by b1rdex** [comment]
Would be nice to see #662 included in 7.0 or some kind of it — I mean the ability to disable `@Inject` thing completely.

**Comment by mnapoli** [comment]
@b1rdex why would you want to disable `@Inject`? Annotations are disabled by default anyway, so it's already the case 🤔

**Comment by b1rdex** [comment]
I thought disabling it has some performance benefits. 

Отправлено с iPhone

**Comment by mnapoli** [comment]
Not really, if you don't use annotations those are not read.

**Comment by juliangut** [comment]
Little feedback on v7

I've created a new branch at https://github.com/juliangut/slim-php-di targetting 7.0-beta2 and so far everything went smoothly with PHP7.4, didn't test in PHP8 yet

**Comment by mnapoli** [comment]
Awesome, thanks for the feedback @juliangut 

I'm using that new version in a prototype for now, going great. PHP 8 attributes are super cool!

**Comment by arokettu** [comment]
A small attribute feedback: unlike phpdoc, attributes can be assigned to parameters so 

```php
    #[Inject(['param2' => 'db.host'])]
    public function method3(Foo $param1, $param2)
    {
    }
```

can be

```php
    public function method3(Foo $param1, #[Inject('db.host')] $param2)
    {
    }
```

It would be nice to have it

**Comment by mnapoli** [comment]
@arokettu absolutely, thank you for pointing it out! I completely forgot about that until I see that Symfony did exactly that recently.

I want to implement that now 👍

**Comment by mnapoli** [comment]
@arokettu I've implemented that in #741

**Comment by arokettu** [comment]
@mnapoli awesome, thanks

File: .php-cs-fixer.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+$finder = PhpCsFixer\Finder::create()
+    ->in(__DIR__ . '/src')
+    ->notPath('Compiler/Template.php');
+$config = new PhpCsFixer\Config;
+
+return $config->setRules([
+    '@PSR2' => true,
```
Comment (SvenRtbg) [inline .php-cs-fixer.php:9]: You probably want PSR12 here

Comment (mnapoli) [reply .php-cs-fixer.php:9]: Not interested in PSR-12 honestly. This is a rename of the file that exists on master.

**Comment by opengeek** [comment]
Is there any chance this will get merged into a deployable branch via composer soon? Even if it is not stable, I need support for psr/container ^2. Without it, I have several projects that are dead in the water because of a complex web of dependency conflicts. If there is anything I can do to support this effort, I would be glad to do what I can.

**Comment by mnapoli** [comment]
@opengeek you should be able to use `dev-v7`. Also some beta versions of v7 have been released already, so you can use these too.

**Comment by odan** [comment]
Hi! :-) Thanks for this awesome package!

I just wanted to say that I and a lot of people in the Slim-Framework community looking forward to this release, because since [Slim 4.8](https://github.com/slimphp/Slim/pull/3063), the `psr/container` version 2 will be installed **by default** and causes an error message when installing php-di v6. Downgrading to `psr/container` version 1 helps, but is causes a lot of troubles and support questions.

**Comment by mnapoli** [comment]
Hey a quick update: I think PHP-DI 7 is ready for release.

It will not support PHP 7.x as this is too much effort, and PHP 7 support will be dropped in 4 months.

If you see anything else to bring up before the v7 release, now is the time :)

**Comment by jacekkarczmarczyk** [comment]
> If you see anything else to bring up before the v7 release, now is the time :)

Is there a migration guide to 7.0 that could be checked before the release?

Also looks like image is not updated
image

**Comment by mnapoli** [comment]
@jacekkarczmarczyk here is the migration guide: https://github.com/PHP-DI/PHP-DI/blob/v7/doc/migration/7.0.md

**Comment by cspray** [comment]
FWIW I have a fairly extensive Container test suite that's been utilizing `v7.x-dev` with 8.1 for quite some time and everything has worked great.

**Comment by route33** [comment]
PHP 7.4 reaches its end of life in 10 days. 
I tested this branch quite extensively during the past month without any issues. When it will be merged?

**Comment by geek-merlin** [comment]
Any updates about the status of all this? I'm confused, see #833.

**Comment by mnapoli** [comment]
Hey everyone, I apologize for dropping the ball here. Big major versions like these are scary, and when you work on a project a few hours every month, it's hard to be confident enough to go over that scariness :)

Though I do appreciate a lot the feedback from those that tested the beta versions (and had no issues). That is great to hear!

Let's move this forward! I am setting a release date to next Thursday because that will be my birthday :p 

I'll tag a release candidate tonight, let's do this!

**Comment by mnapoli** [comment]
https://github.com/PHP-DI/PHP-DI/releases/tag/7.0.0-rc1

**Comment by 306VLD** [comment]
Thank you for your reply. Looking forward to this release 😄

**Comment by geek-merlin** [comment]
Yay, thanks a bunch @mnapoli!

**Comment by mnapoli** [comment]
Today is the day 🎉

https://php-di.org/news/23-php-di-7-0-released.html

**Comment by mnapoli** [comment]
Well thank you :)

**Comment by ldebrouwer** [comment]
Happy birthday @mnapoli! And thank you for all you do, it's really appreciated 🎉

**Comment by rikgirbes** [comment]
Awesome! And happy birthday to you! 🎉

**Pull request merged**